[ARM][feat]: Add 4 bit dynamic quantization matmuls & KleidiAI Backend

nikhil-arm opened the pull request:
Description:
1. Quantize Linear Layer Weights to 4-bits:
Quantize the weights of the Linear layer to 4 bits, using symmetric quantization.
Pack two 4-bit weights into one uint8 container.
Choose a quantization scheme (channel-wise or group-wise), with the group size being a multiple of 32.

2. Prepare Quantized Weights, Scales, and Optional Bias:
After quantizing, obtain the quantized_weights, scales, and groupsize.
If the original Linear layer has a bias, prepare it as well.

3. Pack the Weights Efficiently:
Use torch.ops.aten._dyn_quant_pack_4bit_weight to optimally pack the weights, scales, and optional bias.
```python
packed_weights = torch.ops.aten._dyn_quant_pack_4bit_weight(weight, scales_and_zeros, bias, groupsize, in_features, out_features)
```
Input parameters should include:
in_features and out_features (the same as the Linear layer’s corresponding parameters).

4. Perform Dynamic Quantized Matrix Multiplication:
Use torch.ops.aten._dyn_quant_matmul_4bit to perform matrix multiplication with quantized weights.
```python
output = torch.ops.aten._dyn_quant_matmul_4bit(input, packed_weights,  groupsize, in_features, out_features)
```
Inputs required include:
The input tensor, packed_weights , groupsize, and the in_features and out_features.

API Usage: https://github.com/pytorch/pytorch/issues/143289

Model Perf :
7B Transformer model:
Prefill : 340 t/s
Decode  : 40  t/s
2B Transformer model
Prefill : 747 t/s
Decode  : 80  t/s

Tests:
python test/test_linalg.py -k test__dyn_quant_pack_4bit_weight
Ran 1 test in 0.016s

OK

python test/test_linalg.py -k test__dyn_quant_matmul_4bit
Ran 8 tests in 0.077s

OK

python test/test_linalg.py -k test_compile_dyn_quant_matmul_4bit
Ran 8 tests in 11.454s


Change-Id: Ia1672bad5e6ec94e64d8bb1971395d60f4b3a452

Fixes #ISSUE_NUMBER


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @malfet @snadampal @milpuz01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

ng-05 commented:
@pytorchbot label "ciflow/linux-aarch64" "module:arm"

ng-05 commented:
@pytorchbot label "module:arm"

ng-05 commented:
@pytorchbot label "module: arm"

ng-05 commented:
@pytorchbot label "ciflow/linux-aarch64"

malfet commented on aten/src/ATen/Config.h.in:
```diff
@@ -1,3 +1,5 @@
+// Copyright 2024 Arm Ltd. and/or its affiliates
+
```
We usually don't have copyright headers in the individual files, especially in something autogenerated like  `.h.in` ones. I think https://github.com/pytorch/pytorch/pull/126687#issuecomment-2263607610 proposes an idea to add generic copyright header which will avoid need for individual changes like that

ng-05 replied:
ACK and Fixed

malfet commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,311 @@
+#include "kai_kernels.h"
+#include <ATen/Parallel.h>
+
+#include <float.h>
```
```suggestion
#include <cfloat>
```

malfet commented on aten/src/ATen/Context.cpp:
```diff
@@ -352,6 +354,14 @@ bool Context::hasMKLDNN() {
 }
 
+bool Context::hasKleidiAI() {
+#if AT_KLEIDIAI_ENABLED()
+  return true;
+#else
+  return false;
+#endif
```
Nit
```suggestion
  return AT_KLEIDIAI_ENABLED();
```

snadampal commented on aten/src/ATen/Context.cpp:
```diff
@@ -352,6 +354,10 @@ bool Context::hasMKLDNN() {
 }
 
+bool Context::hasKleidiAI() {
+  return AT_KLEIDIAI_ENABLED();
```
given the kernels are pretty much I8MM based, it would be better to check for aarch64 cpus runtime cpuinfo checks or compiler FEATURE flags for I8MM.

ng-05 replied:
Not Required

snadampal commented on test/test_linalg.py:
```diff
@@ -6401,6 +6405,62 @@ def int4_mm(a, b_int32, b_scales_and_zeros):
+    @skipIfNoKleidiAI
+    @onlyCPU
+    @parametrize("m", [1, 32])
+    @parametrize("k", [128, 256])
```
given these quantized kernels are more relevant for LLMs, please add couple of tests with those model sizes as well, may be for 1B and 3B parameter models

ng-05 replied:
Our kernels are not for LLMs. They are general quantized matmul kernels. I reviewed the test cases and kept the dimensions same as other test cases. Just to accomodate your request I will add larger dimensions.

snadampal commented on aten/src/ATen/native/kleidiai/kai_ukernel_interface.h:
```diff
@@ -0,0 +1,145 @@
+#include <unordered_map>
+#if AT_KLEIDIAI_ENABLED()
+
+#include <kai_lhs_quant_pack_qai8dxp_f32.h>
```
why can't these *_interface.cpp and *_interface.h be defined inside the kleidi repo?
these function definitions have nothing to do with torch/aten. 
Moving these to kleidi repo provides a clean interface, which makes its integration into pytorch cleaner,  and also makes it usable in other frameworks.

snadampal replied:
> ML Frameworks need to have adapter to connect with those interfaces as each framework has different API requirements.

But I don't see any PyTorch/aten specifics in any of these interface functions!

snadampal reviewed (changes requested):
thanks for the contribution. please find my comments inline.

ng-05 commented on CMakeLists.txt:
```diff
@@ -390,6 +392,8 @@ cmake_dependent_option(
 cmake_dependent_option(BUILD_BUNDLE_PTXAS "Bundle PTX into torch/bin fodler"
                        OFF "USE_CUDA" OFF)
+cmake_dependent_option(USE_KLEIDIAI "Use KleidiAI for the Arm architecture." ON
+                        "CPU_AARCH64" OFF)
```
KLEIDIAI is internally guarded at compile time and run time with I8MM and DOTPROD flags. This issue can be handled internally so no need to expose it to pytorch?

ng-05 commented on aten/src/ATen/native/kleidiai/kai_pack.h:
```diff
@@ -0,0 +1,86 @@
+    const int64_t n,
+    const int64_t k,
+    const int64_t bl) {
+  // kai supports 32 block size only
```
Yeah the same packing will work.

ng-05 commented on test/test_linalg.py:
```diff
@@ -19,7 +21,8 @@
      make_fullrank_matrices_with_distinct_singular_values,
      freeze_rng_state, IS_ARM64, IS_SANDCASTLE, TEST_OPT_EINSUM, parametrize, skipIfTorchDynamo,
-     setBlasBackendsToDefaultFinally, setLinalgBackendsToDefaultFinally, serialTest)
+     setBlasBackendsToDefaultFinally, setLinalgBackendsToDefaultFinally, serialTest, \
```
Can you point me to the list of CI unit test cases?

ng-05 commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -3606,5 +3612,65 @@ Tensor _int_mm_cpu(const Tensor& self, const Tensor& mat2) {
+    const int64_t k,
+    const int64_t bl) {
+  int64_t pack_bytes = 0;
+#if AT_KLEIDIAI_ENABLED()
```
The kleidiai ops we are expose needs to have their signature in `native_functions.yaml` 
I am not aware of any way to conditionally enable the signature in `native_functions.yaml` . A function signature can not exist in `native_functions.yaml`  without its cpp counterpart. the above mechanism is a way to always have cpp function expose for `native_functions.yaml`  but disable kleidiai call within.

Please suggest a better way if you are aware of it.

ng-05 commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,311 @@
+size_t global_buffer_bytes = 11008 * 4096 * 4;
+uint8_t* lhs_packed =
+    static_cast<uint8_t*>(aligned_alloc(64, global_buffer_bytes));
+
```
ACK

ng-05 commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,311 @@
+// Copyright 2024 Arm Ltd. and/or its affiliates
```
ACK, we will try to accommodate this request. We have dependent changes in torhcchat and torchao that will be affected with namespace change

ng-05 replied:
Actually just checked this again. We dont need the functionality ( Intrinsic specific dispatch ) that native/cpu offers.
The intrinsic specific code is taken care within kleidiai.
We dont see any backend integrated in the native/cpu and its just the cpp kernels. 
Is it okay if we keep the scope limited for now and refactor if need arises?

ng-05 commented on aten/src/ATen/native/kleidiai/kai_pack.h:
```diff
@@ -0,0 +1,86 @@
+  auto& params = kernel.rhs_pack_params;
+  params.lhs_zero_point = 1;
+  params.rhs_zero_point = 8;
+  kernel.kai_run_rhs_pack(
```
We will add the checks in the packing kernels. We would like to avoid checks in inference kernels as it might affect the latency.

ng-05 commented on aten/src/ATen/native/kleidiai/kai_pack.h:
```diff
@@ -0,0 +1,86 @@
+namespace at::native::kleidiai {
+
+template <typename T>
+Tensor kai_pack_rhs_groupwise_int4(
```
we are working towards putting support for them in our future roadmaps. I agree with what you mentioned but we will keep its scope diff than this PRs

ng-05 commented:
@malfet can you please help in merging this PR.

malfet commented:
> @malfet can you please help in merging this PR.

I believe pre-requisite for merging is passing build and test for specific target and this PR clearly fails aarch64 build right now, see https://github.com/pytorch/pytorch/actions/runs/10616020253/job/29425421107?pr=134124 
Perhaps it's just a matter or rebase, but in general, I would strongly advice against merging a relatively large change if one could not get a clear signal from the platform it targets

ng-05 commented:
Hello @malfet ,
We are in the process to refactor the PR after your valuable inputs.
We are planning to :
1. keep all the files , kernel interface and kernel implementation as it is in  aten/src/ATen/native/kleidiai/*
2. Plug kleidiai int4 matmul kernel with _weight_int4pack_mm_cpu() and modify  the signature of _weight_int4pack_mm_cpu() to fit our requirements
3. Plug kleidiai int4 weght pack kernel with _convert_weight_to_int4pack_cpu() and modify the signature of _convert_weight_to_int4pack_cpu() to fit our requirements
4. Register a new op in torchao for kai_pack_rhs_size() kernel and use it in torchao for quantization and packing step
5. Add / Reuse Symmetric_quantization() in torchao . This will be used for functioning of  _weight_int4pack_mm_cpu() and _convert_weight_to_int4pack_cpu() ops
6. Modify/Add the existing tests for _weight_int4pack_mm_cpu() to accomodate kleidiai kernel and use quantization scheme from torchao directly

Please let me know your thoughts on this and if this addresses all your concerns regarding PR.

digantdesai commented on test/test_linalg.py:
```diff
@@ -6355,45 +6382,61 @@ def test__int4_mm(self, device, m, k, n):
-            b_uint8, b_scales_and_zeros = _group_quantize_tensor(
-                b, n_bit=4, q_group_size=q_group
-            )
+            if torch.backends.kleidiai.is_available() and device == "cpu":
+                b_fp32 = b.to(
+                    dtype=torch.float32
+                )  # Kleidiai only supports float32 datatype
+                b_uint8, b_scales_and_zeros = _group_quantize_tensor_symmetric(
+                    b_fp32, n_bit=4, groupsize=q_group, scheme="symmetric_groupwise"
+                )
```
Meta folks,

Looking at this I am wondering, does it make sense for the Kleidi kernels to live in TorchAO? Seems like these kernels are tied to a specific quantization schema defined in TorchAO.

We might not be all set up to accept this in TorchAO, and there is already a precedence of doing this way in ATen but I am curious whether this is the right path long term?

CC: @malfet, @kimishpatel

ng-05 replied:
We had a discussion with @malfet on the complexities. We have a torchao patch in progress and can import the quantization scheme directly from torchao instead of adding _group_quantize_tensor_symmetric() in pytorch.

1. Integrating this change creates a circular dependency, we need to get this change merged first in order to upstream torchao with further changes. The idea is to make this PR self sufficient for aarch64 CI test cases and migrate the quantization schemes to torchao while integrating arm int4 kernels in existing int4_mm op.
2. Do we have source built torchao packages installed on the default CI test environment? is it ideal to import torchao in test_linalg.py?

ng-05 commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,338 @@
+  const bool can_gemm = (m % 8 == 0) ? true : false;
+
+  if (!can_gemm) {
+    id = kai_kernel_id::
```
This is WIP and will be handled within the kleidiAI kernel in future. For now, the checks will handle the edge cases

ng-05 commented:
1. Bazel build is not mature in KleidiAI in current state. We need changes in bazel configuration to make the imports work correctly with relative/absolute paths as per pytorch root. CI that tests bazel build is failing for now.
2. Need approval on backward compatibilty CI checks as discussed in our meeting.

digantdesai commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,336 @@
+                                  m_start + m0, n_start, dst_stride));
+
+    // Quantize and pack the Input
+    kernel_packet.kai_run_lhs_quant_pack(
```
Is it OK to do dynamic quantization of activation for a weight only quant schema/op from accuracy point of view? i.e. why not make it explicit.

ng-05 replied:
The whole idea behind weight only quant is:
1. to quantize weights ahead of time and quantize input during runtime.
2. Leverage hardware int8 inrinsics to speed up matmul . [ (Input { int8 }) ⓧ  (Weight { int4->int8 } ]

Actually, doing dynamic quantisation on the input gives better accuracy as scales and zp are calculated during runtime and are tailored for that particular input as compared to static quantization which pre calculates scales, zp on calibration dataset before hand

ng-05 replied:
What do you mean when you say make it explicit?  Our Kernels literally say "quant_pack_lhs_int4_mm" in their name and we have a comment explaining same in the kai_kernels.h file.
I am open to suggestions

digantdesai replied:
So for weight only quant the compute happens in FP32. I am not arguing for static quant. 

> What do you mean when you say make it explicit? 

Apologies, I should have been more clear. I haven't seen the TorchAO PR for Kleidi yet but what I meant by more explicit is based on what I see in this PR is, if you see the quantization [flavors](https://github.com/pytorch/ao/blob/e283743b3cc4612bb641b88dca3670231724d396/torchao/quantization/quant_api.py#L427-L430), these Kleidi kernels aligns better with `int8_dynamic_activation_int4_weight` as opposed to `int4_weight_only`. The latter uses `torch.ops.aten._weight_int4pack_mm` which you are overloading here.

ng-05 replied:
I understand what you mean now. The initial version of this PR had separate operators for our kernels and did not use 
`torch.ops.aten._weight_int4pack_mm` for multiple reasons along with the one you mentioned. After discussion with Meta engineers, we decided to use _weight_int4pack_mm as the op interface for our kernel

ng-05 replied:
Do we already have a matmul op that uses` int8_dynamic_activation_int4_weight` in torchao or pytorch

digantdesai commented:
> A simple fix was required and there's already a PR in place for this.

Merged, please update this PR to include new cpuinfo. Thanks

albanD commented on test/test_public_bindings.py:
```diff
@@ -133,6 +133,7 @@ def test_no_new_bindings(self):
             "GraphExecutorState",
             "has_cuda",
             "has_cudnn",
+            "has_kleidiai",
```
Can we avoid adding this to the top level `torch.` module and use the `torch.backends.kleidiai.is_available()` directly  please.

ng-05 replied:
The CI fail here with assertion : https://github.com/pytorch/pytorch/blob/83c594ebd6dfa517fdd67ae23929cc60d5fa325d/test/test_public_bindings.py#L255 without this change

We can add the has_kleidiai to : https://github.com/pytorch/pytorch/blob/83c594ebd6dfa517fdd67ae23929cc60d5fa325d/test/test_public_bindings.py#L245 if you feel thats the right way to go

torch.backends.kleidiai.is_available() uses has_kleidiai check internally

ng-05 replied:
https://github.com/pytorch/pytorch/pull/134124/files#diff-0ccbfbf81a5d8848fa3fb17fdb73b88cc65d5004a778d7a131056798e7b80d66R5

malfet commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4139,13 +4139,13 @@
     CUDA: _int_mm_out_cuda
 
-- func: _convert_weight_to_int4pack(Tensor self, int innerKTiles) -> Tensor
+- func: _convert_weight_to_int4pack(Tensor self, int innerKTiles, int qGroupSize, int N, Tensor qScaleAndZeros, Tensor bias) -> Tensor
```
Can you please document what those extra arguments are doing, specifically what is `N` parameter, and aren't `qGroupSize` can be already informed from the shape of `self`

malfet commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4139,13 +4139,13 @@
     MPS: _convert_weight_to_int4pack_mps
 
-- func: _weight_int4pack_mm(Tensor self, Tensor mat2, int qGroupSize, Tensor qScaleAndZeros) -> Tensor
+- func: _weight_int4pack_mm(Tensor self, Tensor mat2, int qGroupSize, Tensor qScaleAndZeros, int N) -> Tensor
```
Please document the semantic of `N` and why it could not be inferenced from the shape of `self`?

ng-05 commented:
We have added a new 4 bit matmul operator along with its reference implementation to work across all CPU architectures.
The new Operator ```_dyn_quant_matmul_4bit_cpu```  dynamically quantizes the input tensor from fp32 to int8 and matmuls it with prepacked 4 bit groupwise weights. A group can be of a full channel as well.

cfRod commented:
Hi @malfet ,
We have also provided the reference kernels here https://github.com/pytorch/pytorch/pull/134124/files#diff-3dcbe3f22985e9eb2b5ce8fbfb3c7222cef9a7ad1de4a44494995c98d4b4f77cR816 and https://github.com/pytorch/pytorch/pull/134124/files#diff-3dcbe3f22985e9eb2b5ce8fbfb3c7222cef9a7ad1de4a44494995c98d4b4f77cR975

digantdesai commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,451 @@
+      matmul_clamp_f32_qai8dxp1x8_qsi4c32p8x8_1x8x32_neon_dotprod;
+  if (m > 1) {
+    id =
+        kai_kernel_id::matmul_clamp_f32_qai8dxp4x8_qsi4c32p4x8_4x8x32_neon_i8mm;
```
can you point me where we are checking if at runtime this cpu can run i8mm?

digantdesai commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -773,10 +778,409 @@ void int4pack_mm_kernel(
+#endif
+}
+
+inline void ref_dyn_quant_matmul_4bit_channelwise_kernel(
```
Actually, thinking more, this is OK TBH.

cfRod commented:
Hi @digantdesai , Thanks for the comment. We are looking into it.

nikhil-arm committed:
[ARM][feat]: Add 4 bit dynamic quantization  matmuls & KleidiAI Backend

Description:
1. Quantize Linear Layer Weights to 4-bits:
Quantize the weights of the Linear layer to 4 bits, using symmetric quantization.
Pack two 4-bit weights into one uint8 container.
Choose a quantization scheme (channel-wise or group-wise), with the group size being a multiple of 32.

2. Prepare Quantized Weights, Scales, and Optional Bias:
After quantizing, obtain the quantized_weights, scales, and groupsize.
If the original Linear layer has a bias, prepare it as well.

3. Pack the Weights Efficiently:
Use torch.ops.aten._dyn_quant_pack_4bit_weight to optimally pack the weights, scales, and optional bias.
packed_weights = torch.ops.aten._dyn_quant_pack_4bit_weight(weight, scales_and_zeros, bias, groupsize, in_features, out_features)
Input parameters should include:
in_features and out_features (the same as the Linear layer’s corresponding parameters).

4. Perform Dynamic Quantized Matrix Multiplication:
Use torch.ops.aten._dyn_quant_matmul_4bit to perform matrix multiplication with quantized weights.
output = torch.ops.aten._dyn_quant_matmul_4bit(input, packed_weights, scales_and_zeros, bias, groupsize, in_features, out_features)
Inputs required include:
The input tensor, packed_weights, scales, bias, groupsize, and the in_features and out_features.

Model Perf :
7B Transformer model:
Prefill : 340 t/s
Decode  : 40  t/s
2B Transformer model
Prefill : 747 t/s
Decode  : 80  t/s

Tests:
python test/test_linalg.py -k test__dyn_quant_pack_4bit_weight
Ran 1 test in 0.016s

OK

python test/test_linalg.py -k test__dyn_quant_matmul_4bit
Ran 8 tests in 0.077s

OK

python test/test_linalg.py -k test_compile_dyn_quant_matmul_4bit
Ran 8 tests in 11.454s

Signed-off-by: Nikhil Gupta <nikhil.gupta2@arm.com>
Change-Id: I0a9c864c56a9d1b4e6179dc3059cd37b11525c8d

digantdesai commented:
> observed big regression in the torch.compile()

What is the baseline? And did you profile where is time being spent?

> "is missing a c-shim implementation, using proxy executor as fallback"

Hmm.. not super familiar with it, but per the comment at the raise site, I guess you already added this op in `torchgen/aoti/fallback_ops.py`?

ng-05 commented:
> > observed big regression in the torch.compile()
> 
> What is the baseline? And did you profile where is time being spent?
> 
> > "is missing a c-shim implementation, using proxy executor as fallback"
> 
> Hmm.. not super familiar with it, but per the comment at the raise site, I guess you already added this op in `torchgen/aoti/fallback_ops.py`?

I profiled end to ende execution of 4 bit matmul op before and after. This data is average of multiple runs.
matmul shape [M=1 , N = 4096 , K = 4096]
`Compile mode`
Before : 129 us 
After :    232 us

`Egare Mode`
Before : 79 us
After :    80 us

I used the torch profiler to get a breakdown. 
The actual 4 bit matmul operation is taking exactly the same time before and after (55 us) but the overall execution of the compiled graph is slower in compile mode in latest pytorch.

ng-05 commented on torch/_meta_registrations.py:
```diff
@@ -3284,6 +3284,129 @@ def meta__weight_int4pack_mm(x, w, q_group_size, q_scale_and_zeros):
+                assert (bl % kai_bl_multiple_of) == 0
+                return (bl // 2) + num_bytes_multiplier_rhs
+
+            return kai_get_rhs_packed_size_rhs_pack_nxk_qsi4c32p_qsu4c32s1s0(
```
The 4 bit kernel that we have integrated are stabilised. There is a risk of it getting out of sync but its low and we ARM can take care of this when we update kleidiai in pytorch.

digantdesai commented:
```
Compile mode
Before : 129 us
After : 232 us
```
> The actual 4 bit matmul operation is taking exactly the same time before and after (55 us) but the overall execution of the compiled graph is slower in compile mode in latest pytorch.

I am not sure how you ran it. Does the profile say where the time outside matmul is spent? Can we amortize any static costs those by running a bunch of times?

cc @malfet might be able to help you with the right way to benchmark things.

ng-05 commented on cmake/Dependencies.cmake:
```diff
@@ -1463,6 +1464,40 @@ if(NOT INTERN_BUILD_MOBILE)
+      endif()
+      set(TEMP_BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS})
+      set(BUILD_SHARED_LIBS OFF CACHE BOOL "Build shared libs" FORCE)
+      set(AT_KLEIDIAI_ENABLED 1)
```
I have added runtime guard for i8mm and dotprod. These are the two major instructions used in our kernels and can protect agains runtime illegal instruction on older hardware.
For building, kleidiai only builds on machine with dotprod and i8mm instructions

ng-05 commented on cmake/Dependencies.cmake:
```diff
@@ -1463,6 +1464,40 @@ if(NOT INTERN_BUILD_MOBILE)
+  if(USE_KLEIDIAI)
+    if(USE_CUDA OR USE_ROCM OR USE_MPS)
+      set(USE_KLEIDIAI OFF CACHE BOOL "Use KLEIDIAI" FORCE)
+      message(STATUS "USE_KLEIDIAI is set to OFF because USE_CUDA or USE_ROCM or USE_MPS is ON")
```
Left over from common op. Thanks, corrected now

ng-05 commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -773,10 +778,409 @@ void int4pack_mm_kernel(
+      packed_weights.sizes() == weights.sizes(),
+      __func__,
+      " : Packed weights shape should be same as weights");
+  packed_weights.copy_(weights);
```
We do not concat multiple tensors to make it explicity for the user via signature. We want to keep the signature more elaboraitve so that it can accommodate other contributors requirements

ng-05 commented on aten/src/ATen/native/cpu/int4mm_kernel.cpp:
```diff
@@ -773,10 +778,409 @@ void int4pack_mm_kernel(
+
+/**
+ * Dynamic Input Quant 4 bit weights matmul execution flow
+              (INT4 Weights + FP scales + FP32 Bias)
```
grpupwise requires bf16 scales and channelwise requires fp32 scales. I am keeping it as it is for now and will put a comment to make it clear

ng-05 commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,451 @@
+    params.scale_dt = kai_datatype::kai_dt_bf16;
+
+    Tensor bf16_scales = scales;
+    // Kleidiai groupwise kernel only accepts bf16 scales
```
Channelwise kernel requires float32 scales and groupwise kernel requires bf16 scales. We will fallback to ref if these conditions are not met and not convert them explicitly

ng-05 commented:
@digantdesai @malfet Please merge this PR. There are two unsuccessful checks but they are not related to this PR IMO.

ng-05 commented on test/test_linalg.py:
```diff
@@ -6532,6 +6536,127 @@ def int4_mm(a, b_int32, b_scales_and_zeros):
+                b_uint8, b_scales_and_zeros, bias, block_size, in_features, out_features)
+            b_int4pack_meta = torch._dyn_quant_pack_4bit_weight(
+                b_uint8, b_scales_and_zeros, bias, block_size, in_features, out_features)
+            self.assertEqual(b_int4pack.shape, b_int4pack_meta.shape)
```
We will take care of this in AO

malfet commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,442 @@
+    }
+  });
+
+  delete[] lhs_packed;
```
It's 2024, all compiler support smart pointers, please use them
```suggestion
```

ng-05 replied:
done

malfet commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,442 @@
+  uint8_t* dst_act_mtx_f32 = reinterpret_cast<uint8_t*>(output.data_ptr());
+  uint8_t* lhs_native_mtx_f32 = reinterpret_cast<uint8_t*>(input.data_ptr());
+  uint8_t* rhs_packed_mtx_qs4cx = reinterpret_cast<uint8_t*>(weight.data_ptr());
+  uint8_t* lhs_packed = new uint8_t[lhs_packed_size * num_threads];
```
Please use managed pointers
```suggestion
  auto lhs_packed = std::make_unique<uint8_t[]>(lhs_packed_size * num_threads);
```

ng-05 replied:
done

malfet commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,442 @@
+  uint8_t* lhs_packed = new uint8_t[lhs_packed_size * num_threads];
+
+  at::parallel_for(0, num_threads, 0, [&](int64_t begin, int64_t end) {
+    for (int64_t i = begin; i < end; ++i) {
```
```suggestion
    for (const auto  i: c10::irange(begin, end)) {
```

ng-05 replied:
done

malfet commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,442 @@
+    }
+  });
+
+  delete[] lhs_packed;
```
Please use managed pointers
```suggestion
```

ng-05 replied:
done

malfet commented:
> @digantdesai @malfet Please merge this PR. There are two unsuccessful checks but they are not related to this PR IMO.

@ng-05 I believe we had this discussion a while back:
 - Why does is have to go into core vs AO?
 - PR title/description are very confusing to the users: this is indeed a new feature, so title should be something like `Add 4 bit dynamic quantization matmuls` and description should mention how to use new API and perf numbers that dynamic quantization delivers on supported platforms vs say static 4 bit quantization

ng-05 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4177,6 +4177,14 @@
+  dispatch:
+    CPU: _dyn_quant_pack_4bit_weight_cpu
+
+- func: _dyn_quant_matmul_4bit(Tensor inp, Tensor packed_weights, Tensor scales_zeros, Tensor? bias, int block_size, int in_features, int out_features) -> Tensor
```
I have replied to you previous comment already. Please check. https://github.com/pytorch/pytorch/pull/134124#discussion_r1858811842

digantdesai replied:
> The data type of scales_zeroes acts a switch between reference and kleidiai implementation during _dyn_quant_matmul_4bit call.

hmm.. you already have this information when you first pack. If you update the reference packing, when we are anyway [deciding](https://github.com/pytorch/pytorch/pull/134124/files#diff-3dcbe3f22985e9eb2b5ce8fbfb3c7222cef9a7ad1de4a44494995c98d4b4f77cR814) the Kleidi compatibility, you don't need these args at matmul time.

The reason I am pushing back is, from an API point of view, it is odd that we are using the supplied argument only sometimes.

ng-05 replied:
How will you transfer the information about scale dtype from pack to matmul if everything is packed in one big tensor of same dtype?  We will still need to pass an extra argument and the probem will stay exactly same.
I am repeating this again, this API decision was NOT done just for reference implementation. I can update the reference packing and that is not an issue. 
But to decide between ref vs kleidiai implementation we need a switch logic or cast the scales dtype intenally as we were doing before.
Moreover this signature allows us to NOT pack scales_zeroes , bias into one packed buffer. 
What If someone wants to re-ararange and pack just weights in `_dyn_quant_pack_4bit_weight` and use scales, zeroes , bias independently at matmul time. 
You can see this is happening even now here: https://github.com/pytorch/pytorch/blob/920e4364b7f7bd48a4da073893dec8995c1902a2/aten/src/ATen/native/LinearAlgebra.cpp#L3461

From an API POV I believe this is a better API then what is suggested as it clearly depicts the data required for a dynamic quantized matmul and keeps future scope open for extension & other contributors.
I dont like the idea of jam-packing mutliple tensor into one big thing even when some arguments are unused conditionally.

This is my take on the API but if you disagree then we will modify this again and cast the scales internally.

ng-05 replied:
Lets try to get this merge before the coming code freeze. We can always modify this later if theres a requirement as this a fresh API and not integrated anywhere yet.

ng-05 commented:
> > @digantdesai @malfet Please merge this PR. There are two unsuccessful checks but they are not related to this PR IMO.
> 
> @ng-05 I believe we had this discussion a while back:
> 
> * Why does is have to go into core vs AO?
> * PR title/description are very confusing to the users: this is indeed a new feature, so title should be something like `Add 4 bit dynamic quantization matmuls` and description should mention how to use new API and perf numbers that dynamic quantization delivers on supported platforms vs say static 4 bit quantization

@milpuz01 @cfRod I think we had an agreement on 1st point with meta team already?

milpuz01 commented:
> > * Why does is have to go into core vs AO?

> @milpuz01 @cfRod I think we had an agreement on 1st point with meta team already?

Yes, when @digantdesai reviewed PR early in November he was happy with new ATen operation for dynamic quantization to go to core as long as there is a reference implementation for setup without KleidiAI which was added by @ng-05 now.

nikhil-arm committed:
[Refactor]: Remove scale, bias dependency from dyn_quant_matmul_4bit API

Description:
1. The scale, bias tensors are no longer needed in the dynamic quantized
   4 bit matmul call.

Signed-off-by: Nikhil Gupta <nikhil.gupta2@arm.com>
Change-Id: Ia34466eea5fefc5780d418a4321fa9b78c142799

digantdesai commented on aten/src/ATen/native/kleidiai/kai_pack.h:
```diff
@@ -0,0 +1,106 @@
+  auto& params = kernel.rhs_pack_params;
+
+  kernel.kai_run_rhs_pack(
+      1,
```
```suggestion
      /*num_groups=*/1,
```

digantdesai reviewed (approved):
Looks good to me. Thanks for all the back and forth. Left some more comments, mainly nits.

I am approving this to indicate I am ok with merging it, but I would wait for @malfet's final review before merging, since he has also been actively reviewing it.

ng-05 commented on aten/src/ATen/native/kleidiai/kai_kernels.cpp:
```diff
@@ -0,0 +1,440 @@
+    for (const auto x : c10::irange(begin, end)) {
+      matmul_groupwise(
+          std::ref(ukernel),
+          m,
```
In our initial testing, the split m version performed less optimally. It requires re-verification as we have moved to a new kernel for groupwise now.

malfet reviewed (commented):
- If we are adding AOTI bindings, let's have test for it (should go into test/inductor/test_aot_inductor.py )
- Please post some link to RFC outlining the API
- As this is all about performance, please post some short benchmark script in PR description that could be used to validate the perf

ng-05 commented:
> * If we are adding AOTI bindings, let's have test for it (should go into test/inductor/test_aot_inductor.py )
> * Please post some link to RFC outlining the API
> * As this is all about performance, please post some short benchmark script in PR description that could be used to validate the perf

Hello @malfet , thanks for your inputs and support.

Can you please support us on below point,
We had to put our ops in AOTI because of the issue mentioned here :
https://github.com/pytorch/pytorch/pull/134124#issuecomment-2491402828
https://github.com/pytorch/pytorch/pull/134124#issuecomment-2491857545
https://github.com/pytorch/pytorch/pull/134124#issuecomment-2493586097

We see a perf degradation in compile model without adding our op to `fallback_ops.py`, which in turn automatically generates the AOTI code in `c_shim_cpu.cpp`.

We did not intend to enable AOTI path but because of the above scenario it gets enabled. In my understanding we would need to register a `lowering` function for our ops and only then we can add tests to `test/inductor/test_aot_inductor.py`. 
Is there a way to avoid register lowering? or alternate approach that doesn't put the ops in  `fallback_ops.py` ?

2.) The RFC link is already in the PR. I will add it again.

3.) I will share a perf script on my github and put it as part of PR description.

albanD commented on test/test_public_bindings.py:
```diff
@@ -133,6 +133,7 @@ def test_no_new_bindings(self):
             "GraphExecutorState",
             "has_cuda",
             "has_cudnn",
+            "has_kleidiai",
```
This should be removed now that it is expected to be read from torch.backends?

ng-05 replied:
Hello AlbanD, can you please check this https://github.com/pytorch/pytorch/pull/134124#discussion_r1757305689
and https://github.com/pytorch/pytorch/actions/runs/12355469852/job/34482254051?pr=134124

We have removed "has_kleidiai" from test_public_bindings and the ci is failing  https://github.com/pytorch/pytorch/actions/runs/12355469852/job/34482254051?pr=134124

ng-05 replied:
How can we make this change and pass the CI?

ng-05 replied:
Update: I have again reverted this change unless we get a solution for the CI failure.

ng-05 replied:
Got it, changes are done and pushed

ng-05 commented:
RFC for new APIs and Usage
https://github.com/pytorch/pytorch/issues/143289

nikhil-arm committed:
[Feat]: Add AOTI inductor test case for _dyn_quant_matmul_4bit and _dyn_quant_pack_4bit_weight

Tests:

python test/inductor/test_torchinductor.py -k test__dyn_quant_matmul_4bit

Ran 1 test in 0.326s

OK

python test/inductor/test_torchinductor.py -k test__dyn_quant_pack_4bit_weight

Ran 1 test in 5.664s

OK

Signed-off-by: Nikhil Gupta <nikhil.gupta2@arm.com>
Change-Id: I5ed5d8d761769c1af611f75d05e9fc6e9fc64cb4

malfet commented:
@pytorchbot merge -i

malfet commented:
@pytorchbot revert -m "This broke S390 builds, includes cpuinfo unconditionally" -c nosignal

nikhil-arm committed:
[Fix]: Fix s390 CI failures and guard cpuinfo inclusions more tightly

Signed-off-by: Nikhil Gupta <nikhil.gupta2@arm.com>
Change-Id: I08dcdc5780831771a66325ac5e8b45e0805cf990

ng-05 commented:
> @pytorchbot revert -m "This broke S390 builds, includes cpuinfo unconditionally" -c nosignal

Just fixed a push : https://github.com/pytorch/pytorch/pull/134124/commits/6d178afd7d82d2ed12b7734e7e2b350a2d332e1c

ng-05 commented:
@pytorchbot merge -i

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it breaks lots of internal build' -c ghfirst

The build failure seems to happen on x86_64:

```
Stderr:
fbcode/caffe2/aten/src/ATen/Context.cpp:380:10: error: expected expression
  380 |   return AT_KLEIDIAI_ENABLED();
      |          ^
buck-out/v2/gen/fbcode/c9d4f9318ea0f90b/caffe2/__generated-config-header__/buck-headers/ATen/Config.h:22:31: note: expanded from macro 'AT_KLEIDIAI_ENABLED'
   22 | #define AT_KLEIDIAI_ENABLED() @AT_KLEIDIAI_ENABLED@
      |       
```

cc @malfet in case we need to import this

ng-05 commented:
> @pytorchbot revert -m 'Sorry for reverting your change but it breaks lots of internal build' -c ghfirst
> 
> The build failure seems to happen on x86_64:
> 
> ```
> Stderr:
> fbcode/caffe2/aten/src/ATen/Context.cpp:380:10: error: expected expression
>   380 |   return AT_KLEIDIAI_ENABLED();
>       |          ^
> buck-out/v2/gen/fbcode/c9d4f9318ea0f90b/caffe2/__generated-config-header__/buck-headers/ATen/Config.h:22:31: note: expanded from macro 'AT_KLEIDIAI_ENABLED'
>    22 | #define AT_KLEIDIAI_ENABLED() @AT_KLEIDIAI_ENABLED@
>       |       
> ```
> 
> cc @malfet in case we need to import this

Hello, Can you please give me more context on this? What was the failure? which build system? any error log?

huydhn commented:
@pytorchbot rebase

huydhn commented:
@pytorchbot merge -f 'Diff has been landed internally'
